### PR TITLE
Added function `arrayFold()`

### LINKED ISSUE #34155: `arrayFold` function
See

#21589
#23248
#27270

### PR DESCRIPTION
**Pull request by Lirikl**
Fixes #34155

Implemention of function `arrayFold(x1 ,..., xn, accum -> expression, array1,...,arrayn, init_accum)`, applying a lambda function to a number of array columns, collecting the result in an accumulator. The accumulator can be either constant or a column.

The implementation gets the columns of the N-th elements for every array and applies them sequentially.

Examples:
```sql
SELECT arrayFold(x, acc -> x + acc, [1, 2, 3, 4, 5], toInt64(0)); -- sums the elements of array;
SELECT arrayFold(x + acc.1, x * acc.2), [1, 2, 3, 4, 5], (toInt64(0), toInt64(1)); -- sums and multiplies the elements of array;
SELECT arrayFold(x, acc -> arrayPushFront(ACC, x), [1, 2, 3, 4, 5], emptyArrayInt64()); -- reverses array;
```

### Changelog category :
- New Feature

### Changelog entry :
Add function "arrayFold(x1, ..., xn, accum -> expression, array1, ..., arrayn, init_accum)" which applies a lambda function to multiple arrays of the same cardinality and collects the result in an accumulator.

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
This is an automated comment for commit 1a8846cf000b2c936948a12ab232434d8ea8a5ce with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/49794/1a8846cf000b2c936948a12ab232434d8ea8a5ce/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🔴 failure</td></tr>
</table>

**Review by rschu1ze — COMMENTED** [review]
Thanks a lot! Would you add some docs to docs/en/sql-reference/functions/array-functions.md?

And just out of interest: did you have a look at the previous attempts to add arrayFold? #34155. Asking because there were some performance issues and I am curious how this PR addresses them?

**Comment by Lirikl** [comment]
> Thanks a lot! Would you add some docs to docs/en/sql-reference/functions/array-functions.md?
> 
> And just out of interest: did you have a look at the previous attempts to add arrayFold? #34155. Asking because there were some performance issues and I am curious how this PR addresses them?
Yes, I had a look at previous solution, and implemented function in vectorized fashion as was suggested by alexey-milovidov.
I am approximating this solution as at least 100x and some magnitudes faster that implementation in https://github.com/ClickHouse/ClickHouse/pull/21589 depending on array sizes based on few tests including those added in performance tests.

In a week I am planning to add docs and some tests comparing this functions against specified functions such as arraySum

**Review by rschu1ze — COMMENTED** [review]
Will check the code probably tomorrow. It would be nice if you could add some comments to the code that document the design decisions (so novice readers like me will have an easier time 😄 )

**Comment by Lirikl** [comment]
> Will check the code probably tomorrow. It would be nice if you could add some comments to the code that document the design decisions (so novice readers like me will have an easier time 😄 )

eah, alexey-milovidov had the same comment about the lack of comments, will do

File: src/Functions/array/arrayFold.cpp [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    void getLambdaArgumentTypes(DataTypes & arguments) const override
+    {
+        if (arguments.size() < 3)
+            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, "Function {} needs lambda function, at least one array argument and one accumulator argument.", getName());
+        DataTypes nested_types(arguments.size() - 1);
+        for (size_t i = 0; i < nested_types.size() - 1; ++i)
+        {
+            const DataTypeArray * array_type = checkAndGetDataType<DataTypeArray>(&*arguments[i + 1]);
+            if (!array_type)
+                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Argument {} of function {} must be array. Found {} instead.", toString(i + 2), getName(), arguments[i + 1]->getName());
```
Comment (rschu1ze) [inline src/Functions/array/arrayFold.cpp:38]: (`toString()` is superfluous)

File: src/Functions/array/arrayFold.cpp [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+            nested_types[i] = recursiveRemoveLowCardinality(array_type->getNestedType());
+        }
+        nested_types[nested_types.size() - 1] = arguments[arguments.size() - 1];
```
Comment (rschu1ze) [inline src/Functions/array/arrayFold.cpp:41]: `nested_types.back() = arguments.back()`?

File: src/Functions/array/arrayFold.cpp [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+
+        const DataTypeFunction * function_type = checkAndGetDataType<DataTypeFunction>(arguments[0].get());
+        if (!function_type || function_type->getArgumentTypes().size() != nested_types.size())
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "First argument for this overload of {} must be a function with {} arguments. Found {} instead.",
+                            getName(), toString(nested_types.size()), arguments[0]->getName());
```
Comment (rschu1ze) [inline src/Functions/array/arrayFold.cpp:46]: (toString() isn't necessary)

File: src/Functions/array/arrayFold.cpp [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+        if (! accumulator_type->equals(*lambda_type))
+            throw Exception(ErrorCodes::TYPE_MISMATCH, "Return type of lambda function must be the same as the accumulator type. "
+                            "Inferred type of lambda {}, inferred type of accumulator {}.", lambda_type->getName(), accumulator_type->getName());
+
+        return DataTypePtr(accumulator_type);
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+    {
+        const auto & column_with_type_and_name = arguments[0];
```
Comment (rschu1ze) [inline src/Functions/array/arrayFold.cpp:70]: `lambda_with_type_and_name`? Or just `lambda`?

File: src/Functions/array/arrayFold.cpp [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+
+        if (!column_with_type_and_name.column)
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "First argument for function {} must be a function.", getName());
+
+        const auto * column_function = typeid_cast<const ColumnFunction *>(column_with_type_and_name.column.get());
```
Comment (rschu1ze) [inline src/Functions/array/arrayFold.cpp:75]: `lambda_function`?

File: src/Functions/array/arrayFold.cpp [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+        const ColumnArray * column_first_array = nullptr;
+        ColumnsWithTypeAndName arrays;
+        arrays.reserve(arguments.size() - 1);
+
+        for (size_t i = 1; i < arguments.size() - 1; ++i)
+        {
+            const auto & array_with_type_and_name = arguments[i];
+            ColumnPtr column_array_ptr = array_with_type_and_name.column;
+            const auto * column_array = checkAndGetColumn<ColumnArray>(column_array_ptr.get());
+            const DataTypePtr & array_type_ptr = array_with_type_and_name.type;
```
Comment (rschu1ze) [inline src/Functions/array/arrayFold.cpp:91]: Let's move declarations closer to their point of use. E.g. l. 91 and 92 could be moved before l. 101.

File: src/Functions/array/arrayFold.cpp [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+            const auto * array_type = checkAndGetDataType<DataTypeArray>(array_type_ptr.get());
+            if (!column_array)
+            {
+                const ColumnConst * column_const_array = checkAndGetColumnConst<ColumnArray>(column_array_ptr.get());
+                if (!column_const_array)
+                    throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Expected array column, found {}", column_array_ptr->getName());
+                column_array_ptr = recursiveRemoveLowCardinality(column_const_array->convertToFullColumn());
```
Comment (rschu1ze) [inline src/Functions/array/arrayFold.cpp:98]: It looks like the code to remove low cardinality only runs if the array argument is const (l. 93). Do we need some special treatment for low-cardinality, non-const array arguments?

EDIT: Looks like l. 120 handles it.

File: src/Functions/array/arrayFold.cpp [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+                column_array = checkAndGetColumn<ColumnArray>(column_array_ptr.get());
+            }
+            if (!array_type)
+                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Expected array type, found {}", array_type_ptr->getName());
+            if (!offsets_column)
+            {
```
Comment (rschu1ze) [inline src/Functions/array/arrayFold.cpp:104]: (cosmetic: let's omit parentheses for single-line if statements)

(oh, and feel free to apply the same fixes also to Functions/array/FunctionArrayMapped.h which seems like a source of ~~copy-paste~~ inspiration 😄 )

File: src/Functions/array/arrayFold.cpp [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+                column_first_array = column_array;
+            }
+            arrays.emplace_back(ColumnWithTypeAndName(column_array->getDataPtr(),
+                                                      recursiveRemoveLowCardinality(array_type->getNestedType()),
+                                                      array_with_type_and_name.name));
+        }
+
+        ssize_t rows_count = input_rows_count;
+        ssize_t data_row_count = arrays[0].column->size();
+        auto array_count = arrays.size();
```
Comment (rschu1ze) [inline src/Functions/array/arrayFold.cpp:126]: auto --> size_t

File: src/Functions/array/arrayFold.cpp [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+            {
+                column_first_array_ptr = column_array_ptr;
+                column_first_array = column_array;
+            }
+            arrays.emplace_back(ColumnWithTypeAndName(column_array->getDataPtr(),
+                                                      recursiveRemoveLowCardinality(array_type->getNestedType()),
+                                                      array_with_type_and_name.name));
+        }
+
+        ssize_t rows_count = input_rows_count;
```
Comment (rschu1ze) [inline src/Functions/array/arrayFold.cpp:124]: `size_t` (unsigned?)

File: src/Functions/array/arrayFold.cpp [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+        size_t cur_ind = 0;
+        ssize_t cur_arr = 0;
+
+        if (data_row_count)
+            while (offsets[cur_arr] == 0)
+                ++cur_arr;
+
+        for (ssize_t i = 0; i < data_row_count; ++i)
+        {
+            selector[i] = cur_ind++;
```
Comment (rschu1ze) [inline src/Functions/array/arrayFold.cpp:149]: Not wrong, but for readability, let's make the postincrement a separate statement (that's also consistent with most of the remaining codebase).

File: src/Functions/array/arrayFold.cpp [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+            if (cur_ind > max_array_size)
+                max_array_size = cur_ind;
+            while (cur_arr < rows_count && cur_ind >= offsets[cur_arr] - offsets[cur_arr - 1])
+            {
+                ++cur_arr;
+                cur_ind = 0;
+            }
+        }
+
+        std::vector<MutableColumns> data_arrays;
```
Comment (rschu1ze) [inline src/Functions/array/arrayFold.cpp:159]: As written elsewhere, below code is a bit hard to grasp. Let's add comments.

**Pull request merged**